File: day_03/p1.py
import re

def parse_input(fname: str) -> str:
    content = ''
    with open(fname, 'r') as f:
        content = f.read().rstrip()
    
    return content

def compute(content: str) -> int:
    op_pattern = r"mul\([0-9]{1,3},[0-9]{1,3}\)"
    operands_pattern= r"mul\((\d+),(\d+)\)"

    matched_instructions  = re.findall(op_pattern, content)

    operands = [re.match(operands_pattern, op) for op in matched_instructions]
    tuples = [( int(match.group(1)), int(match.group(2)) ) for match in operands]

    return sum(pair[0] * pair[1] for pair in tuples)

def main():
    input = parse_input("input.txt")
    print(compute(input))

if __name__ == "__main__":
    main()

File: day_03/p1.cpp
#include "pypp.hpp" // https://github.com/A-Bergonzoli/pypp/tree/main

#include <iostream>
#include <numeric>
#include <regex>

using Tuples = std::vector<std::pair<long, long>>;

std::string Parse(const std::string& fname)
{
    return pypp::splitFileLines(fname)[0];
}

long Compute(const std::string& input)
{
    std::regex operator_pattern(R"(mul\([0-9]{1,3},[0-9]{1,3}\))");
    std::regex operands_pattern(R"(mul\((\d+),(\d+)\))");

    std::sregex_iterator it(input.begin(), input.end(), operator_pattern);
    std::sregex_iterator end {};

    Tuples tuples {};

    while (it != end) {
        std::smatch match;
        auto match_str = it->str();
        std::regex_match(match_str, match, operands_pattern);
        tuples.emplace_back(std::stol(match[1].str()), std::stol(match[2].str()));
        ++it;
    }

    return std::accumulate(tuples.begin(), tuples.end(), 0,
        [](long sum, const std::pair<long, long>& p) { return sum + p.first * p.second; });
}

int main()
{
    auto input = Parse("input.txt");
    std::cout << Compute(input) << std::endl;

    return 0;
}
